### ISSUE TITLE
Detecting tests that use dynamic in VS 2015 RC causes exception

### ISSUE THREAD

**Issue by jvandertil** [fixie/fixie#98]
Happens after a build when detecting tests.

Using a dynamic object causes the same error only with 'custom metadata item kind: 5'.
Using the Nancy.Testing library (specifically the Browser class in it) causes the exception below.

Looks related to: https://github.com/Fody/Fody/issues/141

Full log:
```
------ Discover test started ------
Fixie 1.0.0.15
Skipping C:\Users\Jos\Source\Legion\src\Legion.NancyExtensions\Legion.NancyExtensions\bin\Debug\Legion.NancyExtensions.dll because it is not a test assembly.
Fixie 1.0.0.15
Processing C:\Users\Jos\Source\Legion\src\Legion.NancyExtensions\tests\Legion.NancyExtensions.Tests\bin\Debug\Legion.NancyExtensions.Tests.dll
Microsoft.Cci.Pdb.PdbDebugException: Unknown custom metadata item kind: 6
   at Microsoft.Cci.Pdb.PdbFunction.ReadCustomMetadata(BitAccess bits)
   at Microsoft.Cci.Pdb.PdbFunction..ctor(ManProcSym proc, BitAccess bits)
   at Microsoft.Cci.Pdb.PdbFunction.LoadManagedFunctions(BitAccess bits, UInt32 limit, Boolean readStrings)
   at Microsoft.Cci.Pdb.PdbFile.LoadFuncsFromDbiModule(BitAccess bits, DbiModuleInfo info, IntHashTable names, ArrayList funcList, Boolean readStrings, MsfDirectory dir, Dictionary`2 nameIndex, PdbReader reader)
   at Microsoft.Cci.Pdb.PdbFile.LoadFunctions(Stream read, Dictionary`2& tokenToSourceMapping, String& sourceServerData)
   at Mono.Cecil.Pdb.PdbReader.PopulateFunctions()
   at Mono.Cecil.Pdb.PdbReader.ProcessDebugHeader(ImageDebugDirectory directory, Byte[] header)
   at Mono.Cecil.ModuleDefinition.ProcessDebugHeader()
   at Mono.Cecil.ModuleDefinition.ReadSymbols(ISymbolReader reader)
   at Mono.Cecil.ModuleReader.ReadSymbols(ModuleDefinition module, ReaderParameters parameters)
   at Mono.Cecil.ModuleReader.CreateModuleFrom(Image image, ReaderParameters parameters)
   at Mono.Cecil.ModuleDefinition.ReadModule(Stream stream, ReaderParameters parameters)
   at Mono.Cecil.ModuleDefinition.ReadModule(String fileName, ReaderParameters parameters)
   at Fixie.VisualStudio.TestAdapter.SourceLocationProvider.CacheTypes(String assemblyPath)
   at Fixie.VisualStudio.TestAdapter.SourceLocationProvider.TryGetSourceLocation(MethodGroup methodGroup, SourceLocation& sourceLocation)
   at Fixie.VisualStudio.TestAdapter.VsTestDiscoverer.DiscoverTests(IEnumerable`1 sources, IDiscoveryContext discoveryContext, IMessageLogger log, ITestCaseDiscoverySink discoverySink)
========== Discover test finished: 1 found (0:00:00,0999979) ==========
```

**Comment by JakeGinnivan** [comment]
Switching the visual studio extension from Mono.Cecil to https://www.nuget.org/packages/Fody/1.29.0-beta01 will fix this issue

**Comment by JakeGinnivan** [comment]
Actually latest Mono.Cecil may have this fixed. @simoncropp do you know?

**Comment by SimonCropp** [comment]
@JakeGinnivan it is in the main line of cecil has the fix but i dont know if it is the cecil nuget has it. the instance of fody running inside msbuild uses a custom nuget of cecil that is sometime ahead of the main nuget https://www.nuget.org/packages/fodycecil

**Comment by plioi** [comment]
@JakeGinnivan Thanks for recognizing that the issue likely lies in this dependency.

@SimonCropp Thanks for the info.  I'll see if I can reproduce the issue before attempting to update Mono.Cecil to the latest nuget.

**Comment by plioi** [comment]
@jvandertil Can you post a sample test class that reproduces the issue?

**Comment by jvandertil** [comment]
This is as minimal as it gets:
```csharp
    public class ReproTests
    {
        public void Test()
        {
            dynamic abc = "test";
        }
    }
```

Which results in this output:
```
------ Discover test started ------
Fixie 1.0.0.15
Processing C:\Users\Administrator\Documents\Visual Studio 2015\Projects\Fixie98Repro\Fixie98Repro\bin\Debug\Fixie98Repro.dll
Microsoft.Cci.Pdb.PdbDebugException: Unknown custom metadata item kind: 5
   at Microsoft.Cci.Pdb.PdbFunction.ReadCustomMetadata(BitAccess bits)
   at Microsoft.Cci.Pdb.PdbFunction..ctor(ManProcSym proc, BitAccess bits)
   at Microsoft.Cci.Pdb.PdbFunction.LoadManagedFunctions(BitAccess bits, UInt32 limit, Boolean readStrings)
   at Microsoft.Cci.Pdb.PdbFile.LoadFuncsFromDbiModule(BitAccess bits, DbiModuleInfo info, IntHashTable names, ArrayList funcList, Boolean readStrings, MsfDirectory dir, Dictionary`2 nameIndex, PdbReader reader)
   at Microsoft.Cci.Pdb.PdbFile.LoadFunctions(Stream read, Dictionary`2& tokenToSourceMapping, String& sourceServerData)
   at Mono.Cecil.Pdb.PdbReader.PopulateFunctions()
   at Mono.Cecil.Pdb.PdbReader.ProcessDebugHeader(ImageDebugDirectory directory, Byte[] header)
   at Mono.Cecil.ModuleDefinition.ProcessDebugHeader()
   at Mono.Cecil.ModuleDefinition.ReadSymbols(ISymbolReader reader)
   at Mono.Cecil.ModuleReader.ReadSymbols(ModuleDefinition module, ReaderParameters parameters)
   at Mono.Cecil.ModuleReader.CreateModuleFrom(Image image, ReaderParameters parameters)
   at Mono.Cecil.ModuleDefinition.ReadModule(Stream stream, ReaderParameters parameters)
   at Mono.Cecil.ModuleDefinition.ReadModule(String fileName, ReaderParameters parameters)
   at Fixie.VisualStudio.TestAdapter.SourceLocationProvider.CacheTypes(String assemblyPath)
   at Fixie.VisualStudio.TestAdapter.SourceLocationProvider.TryGetSourceLocation(MethodGroup methodGroup, SourceLocation& sourceLocation)
   at Fixie.VisualStudio.TestAdapter.VsTestDiscoverer.DiscoverTests(IEnumerable`1 sources, IDiscoveryContext discoveryContext, IMessageLogger log, ITestCaseDiscoverySink discoverySink)
========== Discover test finished: 1 found (0:00:00,0468772) ==========
```

(I know, the metadata item kind is 5, not 6, but this was the fastest I could reproduce)

**Comment by plioi** [comment]
Hah! Thanks.

**Comment by plioi** [comment]
I've updated to the latest Mono.Cecil, which appears to include the fix.  Please upgrade to 1.0.0.18. https://github.com/fixie/fixie/releases/tag/v1.0.0.18

Visual Studio is a bit unrefined when it comes to upgrading.  I recommend opening your solution, uninstalling Fixie, closing and reopening the solution, rebuilding the solution, and *then* trying to run all the tests.  Otherwise, you'll get some false positive exceptions in the test log window.

**Comment by plioi** [comment]
Fixed by 7476a6298c7969ffe21b44b646ee815516a813a4
